python/ray/serve/_private/deployment_state.py

```diff
@@ -2284,6 +2287,16 @@ def shutdown(self):
         # TODO(jiaodong): Need to add some logic to prevent new replicas
         # from being created once shutdown signal is sent.
 
+    def is_ready_for_shutdown(self) -> bool:
+        """Return whether all deployments are shutdown.
+
+        Check there are no deployment states and no checkpoints.
+        """
+        return (
+            len(self._deployment_states) == 0
+            and self._kv_store.get(CHECKPOINT_KEY) is None
```

Comment:
do we need to fetch the checkpoint from the kv store on every iteration? or can we cache this info in the class instance?

Reply:
I feel the state of kv store can change in between shutdown and not sure if the this key can be put/ delete from outside the deployment state manager. But given those assumptions hold false, I think implementing a cached flag when those the put/ delete operation happens makes sense!

Reply:
Assuming this path is only called during shutdown I'm ok with leaving it as-is to avoid any race conditions.

Reply:
cool, will leave this as is